src/test.rs

```diff
@@ -112,6 +73,80 @@ mod cp6_782 {
         .unwrap();
 
         assert!(verify_proof(&pvk, &proof, &[c]).unwrap());
-        assert!(!verify_proof(&pvk, &proof, &[Fr::zero()]).unwrap());
+        assert!(!verify_proof(&pvk, &proof, &[a]).unwrap());
+    }
+}
+
+fn test_rerandomize<E>()
+where
+    E: PairingEngine,
+{
+    // First create an arbitrary Groth16 in the normal way
+
+    let rng = &mut test_rng();
+
+    let params =
+        generate_random_parameters::<E, _, _>(MySillyCircuit { a: None, b: None }, rng).unwrap();
+
+    let pvk = prepare_verifying_key::<E>(&params.vk);
+
+    let a = E::Fr::rand(rng);
+    let b = E::Fr::rand(rng);
+    let c = a * &b;
+
+    // Create the initial proof
+    let proof1 = create_random_proof(
+        MySillyCircuit {
+            a: Some(a),
+            b: Some(b),
+        },
+        &params,
+        rng,
+    )
+    .unwrap();
+
+    // Rerandomize the proof, then rerandomize that
+    let proof2 = rerandomize_proof(rng, &params.vk, &proof1);
+    let proof3 = rerandomize_proof(rng, &params.vk, &proof2);
+
+    // Check that the proofs are equivalent
+
+    assert!(verify_proof(&pvk, &proof1, &[c]).unwrap());
+    assert!(verify_proof(&pvk, &proof2, &[c]).unwrap());
+    assert!(verify_proof(&pvk, &proof3, &[c]).unwrap());
+
+    assert!(!verify_proof(&pvk, &proof1, &[E::Fr::zero()]).unwrap());
+    assert!(!verify_proof(&pvk, &proof2, &[E::Fr::zero()]).unwrap());
+    assert!(!verify_proof(&pvk, &proof3, &[E::Fr::zero()]).unwrap());
```

Comment:
Can you add a check that none of the re-randomized proofs are the same proof?

Reply:
This maybe even belongs in the rerandomize method, so that it panics if randomization fails.

Reply:
I added it to the test. Why should this check be in the rerandomize method? The chance of an incomplete rerandomization is the chance that r_2 = 0, which is 2^(-255) or so